node/network/approval-distribution/src/lib.rs

```diff
@@ -0,0 +1,839 @@
+	peer_views: HashMap<PeerId, View>,
+}
+
+// TODO: Make it public and put in primitives?
+type CandidateIndex = u32;
+
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+enum MessageFingerprint {
+	Assignment(Hash, CandidateIndex, ValidatorIndex),
+	Approval(Hash, CandidateIndex, ValidatorIndex),
+}
+
+#[derive(Debug, Clone, Default)]
+struct Knowledge {
+	known_messages: HashSet<MessageFingerprint>,
+}
+
+/// Information about blocks in our current view as well as whether peers know of them.
+struct BlockEntry {
+	/// Peers who we know are aware of this block and thus, the candidates within it.
+	/// This maps to their knowledge of messages.
+	known_by: HashMap<PeerId, Knowledge>,
+	/// The number of the block.
+	number: BlockNumber,
+	/// The parent hash of the block.
+	parent_hash: Hash,
+	/// Our knowledge of messages.
+	knowledge: Knowledge,
+	/// A votes entry for each candidate.
+	candidates: HashMap<CandidateIndex, CandidateEntry>,
+}
+
+#[derive(Debug)]
+enum ApprovalState {
+	Assigned(AssignmentCert),
+	Approved(AssignmentCert, ValidatorSignature),
+}
+
+/// Information about candidates in the context of a particular block they are included in.
+/// In other words, multiple `CandidateEntry`s may exist for the same candidate,
+/// if it is included by multiple blocks - this is likely the case when there are forks.
+#[derive(Debug, Default)]
+struct CandidateEntry {
+	approvals: HashMap<ValidatorIndex, ApprovalState>,
+}
+
+#[derive(Debug, Clone)]
+enum MessageSource {
+	Peer(PeerId),
+	Local,
+}
+
+impl MessageSource {
+	fn peer_id(&self) -> Option<PeerId> {
+		match self {
+			Self::Peer(id) => Some(id.clone()),
+			Self::Local => None,
+		}
+	}
+}
+
+impl State {
+	async fn handle_network_msg(
+		&mut self,
+		ctx: &mut impl SubsystemContext<Message = ApprovalDistributionMessage>,
+		metrics: &Metrics,
+		event: NetworkBridgeEvent<protocol_v1::ApprovalDistributionMessage>,
+	) {
+		match event {
+			NetworkBridgeEvent::PeerConnected(peer_id, _role) => {
+				// insert a blank view if none already present
+				self.peer_views.entry(peer_id).or_default();
+			}
+			NetworkBridgeEvent::PeerDisconnected(peer_id) => {
+				self.peer_views.remove(&peer_id);
+				self.blocks.iter_mut().for_each(|(_hash, entry)| {
+					entry.known_by.remove(&peer_id);
+				})
+			}
+			NetworkBridgeEvent::PeerViewChange(peer_id, view) => {
+				self.handle_peer_view_change(ctx, metrics, peer_id, view).await;
+			}
+			NetworkBridgeEvent::OurViewChange(view) => {
+				self.handle_our_view_change(metrics, view).await;
+			}
+			NetworkBridgeEvent::PeerMessage(peer_id, msg) => {
+				self.process_incoming_peer_message(ctx, metrics, peer_id, msg).await;
+			}
+		}
+	}
+
+	async fn handle_new_blocks(
+		&mut self,
+		ctx: &mut impl SubsystemContext<Message = ApprovalDistributionMessage>,
+		metrics: &Metrics,
+		metas: Vec<BlockApprovalMeta>,
+	) {
+		// TODO: get rid of superfluous clones
+		let hashes: HashSet<Hash> = metas.iter().map(|m| m.hash.clone()).collect();
```

Comment:
I expect the allocations within HashSet are far more expensive than the clone of 32 bytes. And that those allocations are actually very cheap.